### PR TITLE
feat: raw tx

### PR DESCRIPTION
**Pull request by karlfloersch**
## Description
Makes L2Geth operate on raw tx data!

### Fixes
- Fixes # [Link to Issue]

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](./CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

**Comment by karlfloersch** [comment]
Maybe we should rename `RawTransaction` to `RawBlob`? or `Raw`? Or idk

**Comment by karlfloersch** [comment]
Also @tynes you might want to add to this PR & delete all the newly dead code in the `sync service`? I assume there is dead code there? Or maybe there isn't? Idk

File: core/types/transaction_meta.go [hunk 1]
```diff
@@ -121,6 +123,14 @@ func TxMetaDecode(input []byte) (*TransactionMeta, error) {
 		meta.QueueIndex = &queueIndex
 	}
 
+	raw, err := common.ReadVarBytes(b, 0, 130000, "RawTransaction")
```
Comment (tynes) [inline core/types/transaction_meta.go:126]: Should we get rid of the concept of max read size? This is `0.13mb` which should be enough but I don't know the consensus tx size limit off of the top of my head

File: rollup/sync_service.go [hunk 2]
```diff
@@ -640,5 +642,92 @@ func (s *SyncService) ApplyTransaction(tx *types.Transaction) error {
 	if err != nil {
 		return fmt.Errorf("invalid transaction: %w", err)
 	}
+
+	// Set the raw transaction data in the meta
+	txRaw, err := getRawTransaction(*tx)
```
Comment (tynes) [inline rollup/sync_service.go]: Assuming the upgrade to use standard RLP serialization, this will be able to be removed

**Comment by tynes** [comment]
`newRPCTransaction` needs updating still
https://github.com/ethereum-optimism/go-ethereum/blob/bfa3007ea4c6b3fb601efb1ecbc86d7dd91d05db/internal/ethapi/api.go#L1201

Update:
This has been addressed in 559e171

File: rollup/sync_service.go (hunk 2 above)
Comment (smartcontracts) [reply rollup/sync_service.go]: I just tried this branch against my WIP encoding PR (https://github.com/ethereum-optimism/contracts/pull/300/files) and replaced `getRawTransaction(*tx)` with `rlp.EncodeToBytes(tx)` and it worked first try. Woot!

**Comment by tynes** [comment]
This results in mismatched state roots. To reproduce:

```bash
$  ./scripts/start.sh --rollup.clienthttp http://127.0.0.1:7878 --verifier --datadir $HOME/.ethereum --rollup.statedumppath https://raw.githubusercontent.com/ethereum-optimism/regenesis/master/mainnet/1.json
```

From https://github.com/ethereum-optimism/scripts
```bash
$ VERIFIER_ENDPOINT=http://127.0.0.1:8545 SEQUENCER_ENDPOINT=https://mainnet.optimism.io node scripts/verifier/compare-verifier-sequencer.js
```
The `compare-verifier-sequencer.js` script will find the first mismatched state root. I think it has to do with the enum for tx type

**Comment by tynes** [comment]
I've updated a branch in the contracts repo based on the contracts that are currently deployed to production to include the fixes required to get this to work.

The branch is found here:
https://github.com/ethereum-optimism/contracts/tree/v0.1.4/enums
From within the repo, run the command:
```bash
$ yarn serve
```

Then when starting geth, use the flag:
```bash
--rollup.statedumppath http://127.0.0.1:8081/state-dump.latest.json
```

Doing a state root check will not be sufficient since the genesis state roots will be different since this involves a change to the contract code of a predeploy. Working towards syncing mainnet and then will do a state diff to see if there are any differences. The only difference should be the codeHash for the sequencer entrypoint

**Comment by tynes** [comment]
With the addition of 3319556 then https://github.com/ethereum-optimism/contracts/pull/303 is not needed to be merged and it will also prevent more difficulties in managing versions and we can move towards https://github.com/ethereum-optimism/contracts/pull/300

**Comment by tynes** [comment]
f2114cb adds a safety check with an error log if a transaction with no raw is detected. This should never happen but will prevent a panic in case it does

**Pull request merged**